File: server/datasets/java_source-code/Multiple-A3/method_extracted_1.java
// Linked List Reversal - DISGUISE: method-extracted
// Derived from organic_1.py. The pointer-swap step inside reverse() has been
// pulled out into a separate helper method swapPointers.

class Node {
    int data;
    Node next;

    public Node(int data) {
        this.data = data;
        this.next = null;
    }
}

class LinkedList {
    Node head;

    public LinkedList() {
        this.head = null;
    }

    public void append(int data) {
        Node newNode = new Node(data);
        if (this.head == null) {
            this.head = newNode;
            return;
        }
        Node current = this.head;
        while (current.next != null) {
            current = current.next;
        }
        current.next = newNode;
    }

    private Node[] swapPointers(Node prev, Node curr) {
        Node nextNode = curr.next;
        curr.next = prev;
        return new Node[]{curr, nextNode};
    }

    public void reverse() {
        Node prev = null;
        Node curr = this.head;
        while (curr != null) {
            Node[] swapped = swapPointers(prev, curr);
            prev = swapped[0];
            curr = swapped[1];
        }
        this.head = prev;
    }

    public void display() {
        Node current = this.head;
        StringBuilder sb = new StringBuilder();
        while (current != null) {
            sb.append(current.data);
            if (current.next != null) {
                sb.append(" -> ");
            }
            current = current.next;
        }
        System.out.println(sb.toString());
    }

    public static void main(String[] args) {
        LinkedList ll = new LinkedList();
        for (int value : new int[]{1, 2, 3, 4, 5}) {
            ll.append(value);
        }
        System.out.println("Original list:");
        ll.display();
        ll.reverse();
        System.out.println("Reversed list:");
        ll.display();
    }
}

File: server/datasets/controlled_benchmarks/python/multi_student_cohorts/cohort_a3/method_extracted_1.py
# Linked List Reversal - DISGUISE: method-extracted
# Derived from organic_1.py. The pointer-swap step inside reverse() has been
# pulled out into a separate helper method _swap_pointers, but the overall
# algorithm and output are identical.

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class LinkedList:
    def __init__(self):
        self.head = None

    def append(self, data):
        new_node = Node(data)
        if not self.head:
            self.head = new_node
            return
        current = self.head
        while current.next:
            current = current.next
        current.next = new_node

    @staticmethod
    def _swap_pointers(curr, prev):
        next_node = curr.next
        curr.next = prev
        return next_node, curr

    def reverse(self):
        prev = None
        curr = self.head
        while curr is not None:
            next_node, prev = self._swap_pointers(curr, prev)
            curr = next_node
        self.head = prev

    def display(self):
        values = []
        current = self.head
        while current:
            values.append(str(current.data))
            current = current.next
        print(" -> ".join(values))


if __name__ == "__main__":
    ll = LinkedList()
    for value in [1, 2, 3, 4, 5]:
        ll.append(value)

    print("Original list:")
    ll.display()

    ll.reverse()

    print("Reversed list:")
    ll.display()
